org.osgi.test.assertj.monitoring/src/main/java/org/osgi/test/assertj/monitoring/internal/MonitoringAssertionImpl.java

```diff
@@ -0,0 +1,297 @@
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED, clazz, map));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilNoMoreServiceEventModifiedWithin(long millis, Class<?> clazz,
+			Map<String, Object> map) {
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED, clazz, map))
+						.and(new NoEventWithinPredicate(millis));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilServiceEventModifiedEndmatch(Class<?> clazz) {
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED_ENDMATCH, clazz));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilNoMoreServiceEventModifiedEndmatchWithin(long millis, Class<?> clazz) {
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED_ENDMATCH, clazz))
+						.and(new NoEventWithinPredicate(millis));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilServiceEventModifiedEndmatch(Class<?> clazz, Map<String, Object> map) {
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED_ENDMATCH, clazz, map));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilNoMoreServiceEventModifiedEndmatchWithin(long millis, Class<?> clazz,
+			Map<String, Object> map) {
+		innerPredicate = hasTypeAnd(ServiceEvent.class,
+				Predicates.ServiceEvents.matches(ServiceEvent.MODIFIED_ENDMATCH, clazz, map))
+						.and(new NoEventWithinPredicate(millis));
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilBundleEvent(Predicate<BundleEvent> predicate) {
+		innerPredicate = hasTypeAnd(BundleEvent.class, predicate);
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilFrameworkEvent(Predicate<FrameworkEvent> predicate) {
+		innerPredicate = hasTypeAnd(FrameworkEvent.class, predicate);
+		return this;
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilNoMoreServiceEventWithin(long timeMs) {
+
+		return untilNoMoreServiceEventWithin(timeMs, (se) -> true);
+	}
+
+	@Override
+	public MonitoringAssertionTimeoutStep untilNoMoreServiceEventWithin(long timeMs, Predicate<ServiceEvent> predicate) {
+
+		innerPredicate = hasTypeAnd(ServiceEvent.class, predicate).and(new NoEventWithinPredicate(timeMs));
+		return this;
+	}
+
+	private class NoEventWithinPredicate implements Predicate<Object> {
+
+		private Runner runner;
+
+		public NoEventWithinPredicate(long timeMs) {
+
+			runner = new Runner(timeMs);
+		}
+
+		@Override
+		public boolean test(Object t) {
+
+			try {
+				return runner.await();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+				return false;
+			}
+
+		}
+	}
+
+	private class Runner {
```

Comment:
I don't understand why this class is called "Runner" - it doesn't actually run anything?